text/0000-offline-buildpackages.md

```diff
@@ -0,0 +1,234 @@
+# Meta
+[meta]: #meta
+- Name: Offline Buildpackages
+- Start Date: 2020-04-27
+- Author(s): Dan Thornton
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+
+# Summary
+[summary]: #summary
+
+One paragraph explanation of the feature.
+
+We need to have a method to attach dependencies to buildpackages to support offline workflows. These dependencies should then be made available via the filesystem during the `build` phase.
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+    - Remove need for two separate `ids` for online vs offline buildpacks
+    - Performance optimizations
```

Comment:
How?

Reply:
Ya this should be 'potential performance optimizations' because it requires some additional lifecycle logic. 
How this 'could' work:
- we have a dependency layer in our buildpackage located at `/cnb/deps/my-org/my-dep`, where the contents of `my-dep` look something like this:
```
├─ /cnb/deps/some-org/my-dep
    ├── bin
    │    └── some-executable
    └── lib
         └── header-file.so
         └── header-file.so
```

This dependency has a `bin` and `lib` folder and looks a lot like a [`<layer>`](https://github.com/buildpacks/spec/blob/master/buildpack.md#build). We could have `<layers>/<layer>` symlink to `/cnb/deps/some-org/my-dep`. The lifecycle would have to add a feature where we add symlinked layers to the final image in order to make use of this symlink. This would let buildpack authors just write a symlink file instead of copying all the `my-dep` bits into a layer. (not too sure about how the lifecycle but guessing this might also save us from creating a new OCI layer for `<layers>/layer>` to be added to final image?)

Reply:
I think it would be valuable if this RFC provided a discussion or proposal of how the lifecycle would be able to pull off this performance optimization. 

It seems that the lifecycle would need access to metadata about where deps could be cross repository blob mounted from. Perhaps, when the builder is created a "deps.toml" could be constructed for the lifecycle that lists where each dep is available on a registry so the lifecycle knows if it should attempt to perform a cross repository blob mount on export. 

Additionally, will the lifecycle create one layer with all deps symlinks or will each symlink be its own layer?

cc @ekcasey

Reply:
Ya, going to table these optimizations to a subsequent rfc